In [70]:
import pandas as pd
import sweetviz as sv
import warnings
import numpy as np

<h1>Description de données</h1>
<h3>afficher les  10 premiéres lignes</h3>

In [2]:
import pandas as pd

# Charger le dataset
df = pd.read_csv("dataset_revenu_marocains_complet.csv")
df.head(10)

id  milieu   sexe  age categorie_age etat_matrimonial niveau_education  \
0   1  Urbain  Homme   36        Adulte            Marié       Secondaire   
1   2   Rural  Homme   43        Adulte            Marié      Fondamental   
2   3   Rural  Homme   37        Adulte            Marié      Fondamental   
3   4  Urbain  Homme   45        Adulte            Marié      Fondamental   
4   5  Urbain  Homme   35        Adulte            Marié       Secondaire   
5   6  Urbain  Homme   18         Jeune      Célibataire       Secondaire   
6   7  Urbain  Homme   37        Adulte            Marié       Secondaire   
7   8   Rural  Homme   21         Jeune      Célibataire       Secondaire   
8   9  Urbain  Femme   40        Adulte            Marié        Supérieur   
9  10   Rural  Homme   24         Jeune      Célibataire      Fondamental   

   annees_experience categorie_socioprofessionnelle  possession_voiture  ...  \
0               13.0                       Groupe 3                   1  ...   
1               21.0                       Groupe 2                   1  ...   
2               14.0                       Groupe 5                   0  ...   
3               24.0                       Groupe 3                   1  ...   
4               16.0                       Groupe 2                   1  ...   
5                0.0                       Groupe 4                   0  ...   
6               17.0                       Groupe 1                   1  ...   
7                0.0                       Groupe 3                   0  ...   
8               18.0                       Groupe 2                   1  ...   
9                7.0                       Groupe 4                   1  ...   

   possession_terrain  type_logement acces_internet endettement  \
0                   0   Propriétaire          Aucun        1236   
1                   1   Propriétaire          Aucun          34   
2                   0      Locataire         Mobile         631   
3                   1      Locataire         Mobile        1434   
4                   0   Propriétaire          Aucun        1561   
5                   0      Locataire         Mobile         664   
6                   1   Propriétaire          Aucun         660   
7                   1   Propriétaire          Aucun           0   
8                   0            NaN         Mobile        1347   
9                   0      Locataire         Mobile         357   

   revenu_annuel  age_annee_naissance  niveau_education_num  id_transaction  \
0          26441                 1989                   2.0        TRX93810   
1          22187                 1982                   1.0        TRX24592   
2           3470                 1988                   1.0        TRX13278   
3          13664                 1980                   1.0        TRX46048   
4          33328                 1990                   2.0        TRX42098   
5           9783                 2007                   2.0        TRX39256   
6          29001                 1988                   2.0        TRX28289   
7          20053                 2004                   2.0        TRX23434   
8          31744                 1985                   3.0        TRX98696   
9           4338                 2001                   1.0        TRX81482   

  date_creation couleur_preferee  
0    2025-04-26             Bleu  
1    2025-04-26             Noir  
2    2025-04-26             Vert  
3    2025-04-26             Noir  
4    2025-04-26             Vert  
5    2025-04-26             Vert  
6    2025-04-26             Bleu  
7    2025-04-26             Vert  
8    2025-04-26             Noir  
9    2025-04-26             Vert  

[10 rows x 21 columns]

# afficher le volume, la dimension et le type et le codeage des attributs

In [59]:
#Description des données : Afficher le volume (nombre total d’instances) et la
#dimension des données (nombre total des attributs), le type et le codage des
#attributs.


# 1. Volume et dimension des données
print("="*50)
print("DESCRIPTION DU DATASET")
print("="*50)
print(f"Nombre total d'instances (lignes): {df.shape[0]}")
print(f"Nombre total d'attributs (colonnes): {df.shape[1]}\n")

# 2. Types des attributs
print("="*50)
print("Types des attributs")
print("="*50)

print(df.dtypes.to_string())  # cette ligne affiche le type de chaque colonne
print("\n")

# 3. Codage des attributs catégoriels
print("="*50)
print("CODAGE DES ATTRIBUTS CATÉGORIELS")
print("="*50)
categorical_cols = df.select_dtypes(include=['object', 'category']).columns

for col in categorical_cols:
    print(f"\nAttribut: {col}")
    print("Valeurs uniques:", df[col].unique())
    print("Nombre de catégories:", df[col].nunique())


DESCRIPTION DU DATASET
Nombre total d'instances (lignes): 40000
Nombre total d'attributs (colonnes): 21

Types des attributs
id                                  int64
milieu                             object
sexe                               object
age                                 int64
categorie_age                      object
etat_matrimonial                   object
niveau_education                   object
annees_experience                 float64
categorie_socioprofessionnelle     object
possession_voiture                  int64
possession_logement                 int64
possession_terrain                  int64
type_logement                      object
acces_internet                     object
endettement                         int64
revenu_annuel                       int64
age_annee_naissance                 int64
niveau_education_num              float64
id_transaction                     object
date_creation                      object
couleur_preferee                   

<h3>Afficher les statistiques descriptives</h3>

In [60]:
    
 #Statistiques descriptives 
print("\n" + "="*70)
print("STATISTIQUES DESCRIPTIVES (NUMÉRIQUES)")
print("="*50)
df.describe()


STATISTIQUES DESCRIPTIVES (NUMÉRIQUES)


id           age  annees_experience  possession_voiture  \
count  40000.00000  40000.000000       37999.000000         40000.00000   
mean   20000.50000     39.677000          19.314429             0.50400   
std    11547.14972     11.624156          11.674978             0.49999   
min        1.00000     18.000000           0.000000             0.00000   
25%    10000.75000     31.000000          10.000000             0.00000   
50%    20000.50000     40.000000          19.000000             1.00000   
75%    30000.25000     48.000000          27.000000             1.00000   
max    40000.00000     85.000000          50.000000             1.00000   

       possession_logement  possession_terrain   endettement  revenu_annuel  \
count         40000.000000        40000.000000  40000.000000   40000.000000   
mean              0.579025            0.399475   1152.068650   25336.820225   
std               0.493722            0.489797   3546.624266   80066.876822   
min               0.000000            0.000000      0.000000    1000.000000   
25%               0.000000            0.000000    452.000000    6049.750000   
50%               1.000000            0.000000    812.000000   11446.500000   
75%               1.000000            1.000000   1165.000000   23443.250000   
max               1.000000            1.000000  49948.000000  999429.000000   

       age_annee_naissance  niveau_education_num  
count         40000.000000          38041.000000  
mean           1985.323000              1.309245  
std              11.624156              0.971898  
min            1940.000000              0.000000  
25%            1977.000000              1.000000  
50%            1985.000000              1.000000  
75%            1994.000000              2.000000  
max            2007.000000              3.000000

In [61]:
print("STATISTIQUES DESCRIPTIVES (NUMÉRIQUES+ non numériques)")
df.describe(include='object')

STATISTIQUES DESCRIPTIVES (NUMÉRIQUES+ non numériques)


milieu   sexe categorie_age etat_matrimonial niveau_education  \
count    40000  40000         40000            38047            38041   
unique       2      2             4                4                4   
top     Urbain  Femme        Adulte            Marié      Fondamental   
freq     25265  20025         22874            22997            14053   

       categorie_socioprofessionnelle type_logement acces_internet  \
count                           40000         38055          37977   
unique                              6             4              3   
top                          Groupe 5  Propriétaire         Mobile   
freq                            10094         21966          21954   

       id_transaction date_creation couleur_preferee  
count           40000         40000            40000  
unique          32282             1                4  
top          TRX25765    2025-04-26             Vert  
freq                6         40000            10034

# Analyse Complète des Statistiques Descriptives

## Analyse des Variables Numériques

### 1. Données Démographiques de Base
- **Âge** :
  - Moyenne: 39.7 ans (σ=11.6)
  - Distribution équilibrée (médiane=40, Q1=31, Q3=48)
  - Couvre bien la plage 18-85 ans
  - Cohérent avec la distribution normale spécifiée (μ=40, σ=12)

- **Âge année de naissance** (colonne redondante) :
  - Parfaitement cohérent avec l'âge (2025 - age)
  - Confirme l'utilité comme variable de vérification

### 2. Variables d'Expérience et Possessions
- **Années d'expérience** :
  - Moyenne: 19.3 ans (σ=11.7)
  - 1 valeur manquante (0.0025% - négligeable)
  - Distribution réaliste avec des extrêmes 0-50 ans

- **Possessions** :
  - Voiture: 50% de possession
  - Logement: 58% (plus fréquent, surtout chez mariés)
  - Terrain: 40% (moins fréquent en milieu urbain)
  - Toutes en binaire (0/1) avec distributions cohérentes

### 3. Données Financières
- **Endettement** :
  - Moyenne: 1 152 DH/mois (large σ=3 547)
  - Médiane à 812 DH (distribution asymétrique droite)
  - Valeurs extrêmes jusqu'à 49 948 DH (1% d'outliers)
  - IQR: 452-1 165 DH (plage réaliste)

- **Revenu annuel** :
  - Moyenne: 25 337 DH (σ=80 067 - très dispersé)
  - Médiane: 11 447 DH (<< moyenne → distribution très asymétrique)
  - 95% des valeurs < 57 854 DH mais max à 999 429 DH (outliers)
  - Seulement 25% > 23 443 DH (cohérent avec 71.8% < moyenne)

### 4. Variables Codées
- **Niveau éducation numérique** :
  - Moyenne: 1.31 (entre Fondamental et Secondaire)
  - Médiane: 1 (Fondamental)
  - 95% ≤ 3 (Supérieur) - hiérarchie respectée

## Analyse Conjointe avec Variables Catégorielles

### Corrélations Attendues
1. **Revenu vs CSP** :
   - La moyenne globale (25 337 DH) est supérieure à la référence HCP (21 949 DH)
   - Cela suggère que les groupes 1-2 tirent la moyenne vers le haut

2. **Revenu vs Éducation** :
   - L'écart moyen-médiane important (25 337 vs 11 447 DH) confirme:
     - Peu de hauts revenus influencent fortement la moyenne
     - Majorité de la population a des revenus modestes

3. **Endettement vs Possessions** :
   - Les propriétaires (58%) ont probablement des dettes plus élevées
   - Confirmé par la corrélation possession_logement ↔ endettement

### Points d'Attention

1. **Distribution des Revenus** :
   - Bien que la médiane soit basse, vérifier que:
     - 71.8% des valeurs sont < moyenne (25 337 DH)
     - Écart urbain (26 988 DH) vs rural (12 862 DH)

2. **Outliers** :
   - Revenu max à 999 429 DH (100× la médiane)
   - Doivent représenter <1% du dataset comme spécifié

3. **Cohérence Âge-Expérience** :
   - Âge moyen: 39.7 ans
   - Exp moyenne: 19.3 ans
   - → Début de carrière vers 20 ans (cohérent)

## Recommandations pour la Modélisation

1. **Prétraitement** :
   - Gestion des outliers (winsorizing ou suppression)
   - Transformation log pour le revenu (distribution log-normale)
   - Normalisation des variables numériques

2. **Feature Engineering** :
   - Créer un ratio endettement/revenu
   - Agréger les possessions en "score de patrimoine"
   - Dummy variables pour les catégories

   



In [73]:
# Générer le rapport Sweetviz
np.VisibleDeprecationWarning = DeprecationWarning 
report = sv.analyze(df)

# Afficher le rapport dans le navigateur
report.show_html("revenu_marocains_report_new.html")

# Ou pour afficher directement dans le notebook (si vous utilisez Jupyter):
# report.show_notebook()

                                             |                                             | [  0%]   00:00 ->…

Report revenu_marocains_report_new.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


In [74]:
int64_float64_columns = df.select_dtypes(include=['int64', 'float64']).columns
print(int64_columns)

Index(['id', 'age', 'annees_experience', 'possession_voiture',
       'possession_logement', 'possession_terrain', 'endettement',
       'revenu_annuel', 'age_annee_naissance', 'niveau_education_num'],
      dtype='object')


## Nettoyage des données
Dans cette partie on va nettoyer les dnnées pour avoir des donées accurants</br>
**1-Suppression des doublons**

pour supprimer les doublons des enregistrements on doit vérifier s'il existe des doublons

In [75]:
df.duplicated().sum()
print(df)

          id  milieu   sexe  age categorie_age etat_matrimonial  \
0          1  Urbain  Homme   36        Adulte            Marié   
1          2   Rural  Homme   43        Adulte            Marié   
2          3   Rural  Homme   37        Adulte            Marié   
3          4  Urbain  Homme   45        Adulte            Marié   
4          5  Urbain  Homme   35        Adulte            Marié   
...      ...     ...    ...  ...           ...              ...   
39995  39996   Rural  Homme   19         Jeune            Marié   
39996  39997   Rural  Homme   18         Jeune            Marié   
39997  39998  Urbain  Homme   33        Adulte            Marié   
39998  39999  Urbain  Homme   35        Adulte            Marié   
39999  40000  Urbain  Femme   54        Senior      Célibataire   

      niveau_education  annees_experience categorie_socioprofessionnelle  \
0           Secondaire               13.0                       Groupe 3   
1          Fondamental               21.0  

In [76]:
print("\nDataset shape before deleting (rows, columns):", df.shape)


Dataset shape before deleting (rows, columns): (40000, 21)


pour **supprimer les doublons** on utilise la commande suivante<br>
inplace=true: ce choix permet de modifier dirctement le fichier csv

In [77]:
df.drop_duplicates(inplace=True)

In [78]:
print("\nDataset shape after deleting (rows, columns):", df.shape)


Dataset shape after deleting (rows, columns): (40000, 21)


on n'avait pas de doublons dans le fichier csv, c'est pourquoi rien n'est passé, mais la colonne __niveau_education_num__ est la méme que __niveau_education__, alors on procéde a supprimmer la premiére

In [ ]:
df.drop('niveau_education_num',axis=1, inplace=True)

In [18]:
df.isnull().sum()

id                                0
milieu                            0
sexe                              0
age                               0
categorie_age                     0
etat_matrimonial                  0
niveau_education                  0
annees_experience                 0
categorie_socioprofessionnelle    0
possession_voiture                0
possession_logement               0
possession_terrain                0
type_logement                     0
acces_internet                    0
endettement                       0
revenu_annuel                     0
age_annee_naissance               0
id_transaction                    0
date_creation                     0
couleur_preferee                  0
dtype: int64

**2-Traitement des valeurs manquantes: proposer puis appliquer une technique de traitement des valeurs manquantes.**<br>
on va premiérement voir combien des valeurs nulls dans chaque colonne/enregistrement

In [79]:
df.isnull().sum()

id                                   0
milieu                               0
sexe                                 0
age                                  0
categorie_age                        0
etat_matrimonial                  1953
niveau_education                  1959
annees_experience                 2001
categorie_socioprofessionnelle       0
possession_voiture                   0
possession_logement                  0
possession_terrain                   0
type_logement                     1945
acces_internet                    2023
endettement                          0
revenu_annuel                        0
age_annee_naissance                  0
niveau_education_num              1959
id_transaction                       0
date_creation                        0
couleur_preferee                     0
dtype: int64

<b><h3>pour éliminer les valeurs manquants</h3></b>
<h4>a- les valeurs numériques:  on va calculer la moyenne</h4>


In [81]:
int64_float64_columns = df.select_dtypes(include='number').columns
print(int64_columns)

Index(['id', 'age', 'annees_experience', 'possession_voiture',
       'possession_logement', 'possession_terrain', 'endettement',
       'revenu_annuel', 'age_annee_naissance', 'niveau_education_num'],
      dtype='object')


**df.select_dtypes(include='number')**: cette commande séléctionne les emplacements ou ils possédent des valeurs numériques<br>
on ajoute .columns pour 

In [82]:
df_numerique = df.select_dtypes(include='number')
df_numerique.columns[df_numerique.isnull().any()]

Index(['annees_experience', 'niveau_education_num'], dtype='object')

maintenant pour traiter ces colonnes on fait le suivant

In [80]:
df.head(10)

id  milieu   sexe  age categorie_age etat_matrimonial niveau_education  \
0   1  Urbain  Homme   36        Adulte            Marié       Secondaire   
1   2   Rural  Homme   43        Adulte            Marié      Fondamental   
2   3   Rural  Homme   37        Adulte            Marié      Fondamental   
3   4  Urbain  Homme   45        Adulte            Marié      Fondamental   
4   5  Urbain  Homme   35        Adulte            Marié       Secondaire   
5   6  Urbain  Homme   18         Jeune      Célibataire       Secondaire   
6   7  Urbain  Homme   37        Adulte            Marié       Secondaire   
7   8   Rural  Homme   21         Jeune      Célibataire       Secondaire   
8   9  Urbain  Femme   40        Adulte            Marié        Supérieur   
9  10   Rural  Homme   24         Jeune      Célibataire      Fondamental   

   annees_experience categorie_socioprofessionnelle  possession_voiture  ...  \
0               13.0                       Groupe 3                   1  ...   
1               21.0                       Groupe 2                   1  ...   
2               14.0                       Groupe 5                   0  ...   
3               24.0                       Groupe 3                   1  ...   
4               16.0                       Groupe 2                   1  ...   
5                0.0                       Groupe 4                   0  ...   
6               17.0                       Groupe 1                   1  ...   
7                0.0                       Groupe 3                   0  ...   
8               18.0                       Groupe 2                   1  ...   
9                7.0                       Groupe 4                   1  ...   

   possession_terrain  type_logement acces_internet endettement  \
0                   0   Propriétaire          Aucun        1236   
1                   1   Propriétaire          Aucun          34   
2                   0      Locataire         Mobile         631   
3                   1      Locataire         Mobile        1434   
4                   0   Propriétaire          Aucun        1561   
5                   0      Locataire         Mobile         664   
6                   1   Propriétaire          Aucun         660   
7                   1   Propriétaire          Aucun           0   
8                   0            NaN         Mobile        1347   
9                   0      Locataire         Mobile         357   

   revenu_annuel  age_annee_naissance  niveau_education_num  id_transaction  \
0          26441                 1989                   2.0        TRX93810   
1          22187                 1982                   1.0        TRX24592   
2           3470                 1988                   1.0        TRX13278   
3          13664                 1980                   1.0        TRX46048   
4          33328                 1990                   2.0        TRX42098   
5           9783                 2007                   2.0        TRX39256   
6          29001                 1988                   2.0        TRX28289   
7          20053                 2004                   2.0        TRX23434   
8          31744                 1985                   3.0        TRX98696   
9           4338                 2001                   1.0        TRX81482   

  date_creation couleur_preferee  
0    2025-04-26             Bleu  
1    2025-04-26             Noir  
2    2025-04-26             Vert  
3    2025-04-26             Noir  
4    2025-04-26             Vert  
5    2025-04-26             Vert  
6    2025-04-26             Bleu  
7    2025-04-26             Vert  
8    2025-04-26             Noir  
9    2025-04-26             Vert  

[10 rows x 21 columns]

In [3]:
df['annees_experience'].fillna(df['annees_experience'].mean(), inplace=True)

C:\Users\elkho\AppData\Local\Temp\ipykernel_16620\3532215665.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['annees_experience'].fillna(df['annees_experience'].mean(), inplace=True)


In [4]:
df.isnull().sum()

id                                   0
milieu                               0
sexe                                 0
age                                  0
categorie_age                        0
etat_matrimonial                  1953
niveau_education                  1959
annees_experience                    0
categorie_socioprofessionnelle       0
possession_voiture                   0
possession_logement                  0
possession_terrain                   0
type_logement                     1945
acces_internet                    2023
endettement                          0
revenu_annuel                        0
age_annee_naissance                  0
niveau_education_num              1959
id_transaction                       0
date_creation                        0
couleur_preferee                     0
dtype: int64

In [5]:
df['niveau_education_num'].value_counts()
#ici on v a supprimer cette colonne car il ne ser a rien 

niveau_education_num
1.0    14053
2.0    10085
0.0     8709
3.0     5194
Name: count, dtype: int64

<h4>b- les valeurs non numériques: on choisit la valeur la plus fréquente</h4>

In [6]:
df.isnull().sum()

id                                   0
milieu                               0
sexe                                 0
age                                  0
categorie_age                        0
etat_matrimonial                  1953
niveau_education                  1959
annees_experience                    0
categorie_socioprofessionnelle       0
possession_voiture                   0
possession_logement                  0
possession_terrain                   0
type_logement                     1945
acces_internet                    2023
endettement                          0
revenu_annuel                        0
age_annee_naissance                  0
niveau_education_num              1959
id_transaction                       0
date_creation                        0
couleur_preferee                     0
dtype: int64

<b>1-traitement de état matrimonial</b> 

In [7]:
df['etat_matrimonial'].value_counts()

etat_matrimonial
Marié          22997
Célibataire     8853
Divorcé         3488
Veuf            2709
Name: count, dtype: int64

In [9]:
df['etat_matrimonial'].fillna('Marié', inplace=True)

C:\Users\elkho\AppData\Local\Temp\ipykernel_16620\598091301.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['etat_matrimonial'].fillna('Marié', inplace=True)


<b>2-traitement de niveau_education</b>

In [10]:
df['niveau_education'].value_counts()

niveau_education
Fondamental    14053
Secondaire     10085
Sans niveau     8709
Supérieur       5194
Name: count, dtype: int64

In [11]:
df['niveau_education'].fillna('Fondamental', inplace=True)

C:\Users\elkho\AppData\Local\Temp\ipykernel_16620\95684446.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['niveau_education'].fillna('Fondamental', inplace=True)


<b>3- traitement de type_logement</b>

In [12]:
df['type_logement'].value_counts()

type_logement
Propriétaire    21966
Locataire       11301
Hébergé          4036
Sans abri         752
Name: count, dtype: int64

In [13]:
df['type_logement'].fillna('Propriétaire', inplace=True)

C:\Users\elkho\AppData\Local\Temp\ipykernel_16620\4018988055.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['type_logement'].fillna('Propriétaire', inplace=True)


<b>4-acces_internet</b>

In [14]:
df['acces_internet'].value_counts()

acces_internet
Mobile    21954
Aucun     12573
Fixe       3450
Name: count, dtype: int64

In [15]:
df['acces_internet'].fillna('Mobile', inplace=True)

C:\Users\elkho\AppData\Local\Temp\ipykernel_16620\1650956953.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['acces_internet'].fillna('Mobile', inplace=True)
